src/cuda/ptx.cu

```diff
@@ -0,0 +1,262 @@
+#include "cuda_runtime.h"
+#include "device_launch_parameters.h"
+
+#include <stdio.h>
+#include <iostream>
+#include <bitset>
+#include <string>
+
+#include "encode_ptx.h"
+
+#define NUM_THREADS 256
+#define NUM_BLOCKS 1024
+#define EIGHT_GB_IN_BYTES 8589934592
+
+#define CUDA_FATAL(expr) do {				\
+    cudaError_t code = expr;				\
+    if (code != cudaSuccess) {				\
+        cerr << #expr << "@" << __LINE__ << " failed: "	\
+             << cudaGetErrorString(code) << endl;	\
+	exit(1);					\
+    }							\
+} while(0)
+
+
+extern "C" bool detect_cuda()
+{
+    cudaDeviceProp prop;
+    return cudaGetDeviceProperties(&prop, 0) == cudaSuccess;
+}
+
+extern "C" bool batch_encode(unsigned int piece[], size_t len,
```

Comment:
I'd like to see some more descriptive name, like `sloth256_189_cuda_encode`

Reply:
how about cuda_batch_encode? 
I want to include `batch` into the name, since this is not able to encode a single piece. 
Wouldn't `sloth256_189_cuda_batch_encode` be too long? And also, isn't it obvious from context that it is sloth256_189? Decision is yours :)

Reply:
`sloth256_189_cuda_batch_encode` is fine, This is C, which means namespace for functions is global, which is why it is a good practice to prefix it with project name or whatnot.